pkg/check/precheck.go

```diff
@@ -0,0 +1,126 @@
+package check
+
+import (
+	"sync"
+
+	"github.com/juju/errors"
+	"github.com/ngaut/log"
+)
+
+// Checker is interface that defines checker.
+type Checker interface {
+	Name() string
+	Check() *Result
+}
+
+// State is state of check
+type State string
+
+const (
+	// StateSuccess indicates that the check was successful
+	StateSuccess State = "success"
+	// StateFailure indicates that the check was faiful
+	StateFailure State = "fail"
+	// StateWarning indicates that the check had warnings
+	StateWarning State = "warn"
+)
+
+// Result is result of check
+type Result struct {
+	ID          uint64 `json:"id"`
+	Name        string `json:"name"`
+	Desc        string `json:"desc"`
+	State       State  `json:"state"`
+	ErrorMsg    string `json:"errorMsg"`
+	Instruction string `json:"instruction"`
+	Extra       string `json:"extra"`
+}
+
+// ResultSummary is summary of all check results
+type ResultSummary struct {
+	Passed     bool  `json:"passed"`
+	Total      int64 `json:"total"`
+	Successful int64 `json:"successful"`
+	Failed     int64 `json:"failed"`
+	Warning    int64 `son:"warning"`
+}
+
+// Results contains all check results and summary
+type Results struct {
+	Results []*Result      `json:"results"`
+	Summary *ResultSummary `json:"summary"`
+}
+
+// Do executes several checkers.
+func Do(checkers []Checker) (*Results, error) {
```

Comment:
do we need to add `ctx` to suppoer cancel

Reply:
good catch